### ISSUE TITLE
In-place operations (dropout, ReLU, etc..)

### ISSUE THREAD

**Issue by alexgkendall** [tensorflow/tensorflow#4309]
As far as I understand, tensorflow does not support in-place operation of operations such as ReLU, Dropout, etc..

If an operation outputs the same type and shape tensor and does not involve any branching in the graph then it should be possible to operate on the same chunk of memory.

As it stands, tensorflow doesn't support this which severely increases memory usage for large deep learning models.

**Comment by martinwicke** [comment]
Since most of the memory is concentrated in the weights, I don't think this would lead to significant savings. The examples you give only affect activations, which are a tiny part of a typical model's memory consumption.

**Comment by vrv** [comment]
Martin: I wouldn't say that all the memory is always concentrated in weights.

However, we have not seen proof that you need in-place operations to be memory efficient.

* If there is only one consumer of an activation, the memory will be freed almost instantaneously after the consuming op is completed.  Yes, you can avoid this one memory allocation for an activation buffer, but in practice this is never the dominant memory cost of a model (One case where this would be practically important is if the activation memory of one layer was the majority of the GPU memory, but I don't believe that's common.)

* As some external studies have shown, TensorFlow is already quite efficient when compared to many other frameworks in using GPU memory already, and I'd like to see proof that in-place updates would help dramatically.  Until then, I think we'd rather spend time on the potential bigger wins in placement and scheduling before trying to implement in-place updates.

**Comment by alexgkendall** [comment]
Thank you for your reply. Consider a 3d convolutional model (say for example over a video tensor - these are quite commonly used). @martinwicke in this situation the dominant memory by far is in the feature tensors. Having to create deep copies of these tensors for both activation and dropout severely limits the depth of networks you can train. This contrasts with other libraries (e.g. Caffe/ Torch) which allow in-place operation.

**Comment by mbrubake** [comment]
This is also true with some fully convolutional models and generative architectures.

**Comment by vrv** [comment]
https://github.com/tensorflow/tensorflow/blob/52dcb2590bb9274262656c958c105cb5e5cc1300/tensorflow/core/framework/op_kernel.h#L714 this has now been implemented and used where we have discovered it is safe to do so.
